### DISCUSSION TITLE
iCub simulator in PyBullet

### DISCUSSION THREAD

**Discussion by rustlluk** [robotology/community#682] [Show and tell]
Hello everyone,
With @matejhof we have a course on humanoid robotics at our university. So far, we used simulated iCub in Gazebo as a platform for assignments. However, our students were complaining about C++, the necessity to learn YARP only for the sake of the course, and the hardware requirements to run the simulation smoothly. 
So, I decided to create a less demanding simulator of iCub in Python based on PyBullet. You can check it here [https://github.com/rustlluk/pyCub](https://github.com/rustlluk/pyCub). 
 I know that [Pybullet-robot-envs](https://github.com/hsp-iit/pybullet-robot-envs) exists, but as it is no longer in development, I decided that something I develop from scratch will be easier to fix and maintain.

An example of a ball on a table scene :
![pycub](https://github.com/robotology/community/assets/54062527/fd9c4214-b845-4255-957c-a508a6023fd6)

What is currently implemented:
- control in joint and cartesian space
  - **without a proper planner**, i.e., users have to take care of the trajectory and collision avoidance
- worlds can be changed through yaml configuration files
- skin can be simulated as well
- visualization is done with Open3D, because I don't like the default one
  - output from both eyes can be accessed as well

It may still be quite buggy. We will start the new semester in mid-February, and I assume the students will test it properly :) My plan is to fix possible problems that may occur during the course, refactor the code after that, and optimize some things. 

I do not have any plans to develop more features, as it is probably enough for our case right now.  However, if someone finds it useful, I am open to new ideas and requests for features :)

**Comment by Nicogene** [comment]
cc @robotology/everyone

**Comment by pattacini** [comment]
Nice work @rustlluk @matejhof 🚀

**Reply by pattacini** [reply]
A side note 😄 

>However, our students were complaining about C++

Today's students!

**Reply by alecive** [reply]
Don't even get me started on this 🤦

**Reply by myl1ne** [reply]
I'm happy to see we all share the same burden xD
